### PR TITLE
Feature specifications to pydantic objects

### LINKED ISSUE #82: refactor: add FeatureSpec and FeatureGroupSpec to prep for JOSS
Use Pydantic base objects for this.

### PR DESCRIPTION
**Pull request by MartinBernstorff**
- [X] I have battle-tested on Overtaci (RMAPPS1279)
- [X] I have assigned ranges (e.g. `>=0.1, <0.2`) to all new dependencies (allows dependabot to keep dependency ranges wide for better compatibility)

Fixes #82.

## Notes for reviewers
This is a big one! Turned out to be **much** harder and more thorough than expected. Our main advantages are:

1. Conceptual definitions of our specifications. We now clearly delineate StaticSpecs from TemporalSpecs.
2. Instant feedback if features are misspecified.
3. Splitting the loading of dataframes out of flattened dataset. Means that FlattenedDataset takes specs with dfs, and others can use whatever loading method they want. Makes it closer to ready for JOSS, since it now only flattens values (and handles cache, which we probably want to move out of it as well).
4. Much faster concatenation - was getting super tired of it while debugging, so changed it so that each subprocess sets and sorts by index. Makes concat 20x+ faster than our previous merge, and probably scales much better as well.

I'd focus on:
* unresolved_feature_spec_objects.py
* feature_spec_objects.py

My primary doubts are re: ineritance - we have about 3 levels of it in `feature_spec_objects` making the code harder to read and edit. If we could use composition for some of it, that should decouple us more.

### CODE REVIEW

File: src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    prefix: str = "outc"
+
+    incident: bool
+    # Whether the outcome is incident or not, i.e. whether you can experience it more than once.
+    # For example, type 2 diabetes is incident. Incident outcomes cna be handled in a vectorised
+    # way during resolution, which is faster than non-incident outcomes.
+
+    def is_dichotomous(self) -> bool:
+        """Check if the outcome is dichotomous."""
+        return len(self.values_df["value"].unique()) <= 2
```
Comment (MartinBernstorff) [inline src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py:151]:
```suggestion
        col_name = "value" if not input_col_name_override else self.input_col_name_override
        return len(self.values_df[col_name].unique()) <= 2
```

**Review by HLasse — APPROVED** [review]
Amazing job, Martin! This is an absolutely massive amount of high-quality code in a very short period of time! I'm sure this will make all of our lives easier down the road- both in application and also for future development. 

There's just a few minor comments scattered throughout.

I'm unsure if the UnresolvedSpecs should be in `application` or not - depends on whether we want to encourage users/give them the opportunity to use registries to load their data. It is quite handy in many cases, but it does make the interface a bit more difficult. We can always discuss this later.

**Comment by MartinBernstorff** [comment]
> Amazing job, Martin! This is an absolutely massive amount of high-quality code in a very short period of time! I'm sure this will make all of our lives easier down the road- both in application and also for future development.
> 
> There's just a few minor comments scattered throughout.
> 
> I'm unsure if the UnresolvedSpecs should be in `application` or not - depends on whether we want to encourage users/give them the opportunity to use registries to load their data. It is quite handy in many cases, but it does make the interface a bit more difficult. We can always discuss this later.

Thanks a ton, Lasse! I agree, it might be useful to have a "load_dfs" or similar folder containing the `unresolved` objects, and abstract away the loading. Looking at your comments now :-)

File: .gitignore [hunk 2]
```diff
@@ -160,4 +160,7 @@ pytest-coverage.txt
 *test_load_data.py
 wandb
 poetry.lock
-*.html
\ No newline at end of file
+*.html
+
+# PyCharm
+/.idea/*
```
[the remark this answers is not in the document]
Comment (MartinBernstorff) [reply .gitignore:166]: Haha, yeah, trying out pycharm! Their refactoring tools are **awesome** - e.g. you can move a function and have all the references automatically updated.

File: src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
```
[the remark this answers is not in the document]
Comment (MartinBernstorff) [reply src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py:152]: Omg, using super here is awesome.

File: src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+from psycop_feature_generation.timeseriesflattener.resolve_multiple_functions import (
+    resolve_multiple_fns,
+)
+from psycop_feature_generation.utils import data_loaders
+
+
+@lru_cache()
+def load_df_with_cache(loader_fn: Callable, kwargs: dict[str, Any]) -> pd.DataFrame:
+    return loader_fn(**kwargs)
```
Comment (KennethEnevoldsen) [inline src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py:21]: You don't have to use the @ you can also do it in the function:

```
memoize_loader_fn = lru_cache(loader_fn)
```
the name is not misspelt it is a term: https://en.wikipedia.org/wiki/Memoization

File: src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+
+    loader_kwargs: Optional[dict[str, Any]] = None
+    # Optional kwargs for the values_loader
+
+    values_df: Optional[pd.DataFrame] = None
+    # Dataframe with the values.
+
+    feature_name: str
+    prefix: str
+    # Used for column name generation, e.g. <prefix>_<feature_name>.
```
Comment (KennethEnevoldsen) [inline src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py:87]: not a fan of these, but it is not too big of a problem.

File: src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+            self.values_df.rename(
+                columns={input_col_name: self.output_col_name_override},
+                inplace=True,
+            )
+
+    def __eq__(self, other):
+        # Trying to run `spec in list_of_specs` works for all attributes except for df,
+        # since the truth value of a dataframe is ambiguous. To remedy this, we use pandas'
+        # .equals() method for comparing the dfs, and get the combined truth value.
```
Comment (KennethEnevoldsen) [inline src/psycop_feature_generation/timeseriesflattener/feature_spec_objects.py:120]: docstring?

File: src/application/t2d/generate_features_and_write_to_disk.py [hunk 7]
```diff
@@ -478,118 +531,38 @@ def main(
...
-
-    predictor_spec_list += generate_feature_specification(
-        dfs=("weight_in_kg", "height_in_cm", "bmi"),
-        lookbehind_days=lookbehind_days,
-        resolve_multiple=["latest"],
-        fallback=np.nan,
-    )
-
-    return predictor_spec_list
+    finish_wandb()
```
Comment (KennethEnevoldsen) [inline src/application/t2d/generate_features_and_write_to_disk.py:561]: not sure if this is needed?

**Review by KennethEnevoldsen — APPROVED** [review]
Looks good to me. A few minor comments.

I haven't looked at the test_*

Also a few large comment block in the application scripts, but it is the application script so 🤷

**Pull request merged**